### PR TITLE
Implement Stripe functionality for checkout

### PR DESCRIPTION
**Pull request by AmberWilkie**
PT Story: https://www.pivotaltracker.com/story/show/131625489

Changes proposed in this pull request:
- Allow customers to check out with Stripe credit card payment system
- Replace old "Pay Now" button with JS Stripe Pay with Card button
- Mock Stripe API calls in test
- Add `stripe_customer` and `paid` columns for `ShoppingCart` - the first is filled with the `id` that comes back on the Stripe JSON. The other is simply flipped when the charge succeeds.

What I have learned working on this feature:
- (Anticipating I will learn) how to mock API calls from Stripe
- How to test Javascript buttons and elements
- Manipulating Stripe callbacks

Screenshots: 

screenshot_2016-10-06-20-35-45 048

Ready for review!

### CODE REVIEW

File: app/controllers/carts_controller.rb [hunk 1]
```diff
@@ -11,12 +11,22 @@ def add_item
...
-    flash[:notice] = 'Your food is on its way!'
-    # In a later feature this needs to create some action item to actually make the order happen.
-    # Restrict this function to a Customer
+  def create
+    cart = ShoppingCart.find(params[:cart_id])
+    cart.user_id = current_user.id
+    cart.save
+    charge = StripePayment.perform_payment(params, cart)
+    @order = ShoppingCart.find_by(user_id: current_user.id)
+    if charge.class == Stripe::Charge
```
Comment (AmberWilkie) [inline app/controllers/carts_controller.rb:18]: This should probably be extracted to a private method for readability

File: features/step_definitions/cart_steps.rb [hunk 2]
```diff
@@ -34,9 +34,59 @@
...
+  sleep(1)
+end
+
+When(/^I click the "([^"]*)" stripe button$/) do |arg|
+  find('.stripe-button-el').trigger('click')
+end
+
+And(/^I wait for Stripe to respond$/) do
+  loop until all(:xpath, '//input[contains(@name, "stripeToken")]').length == 1
+end
```
Comment (AmberWilkie) [inline features/step_definitions/cart_steps.rb]: I don't believe we use this step anymore.

**Comment by AmberWilkie** [comment]
I'd appreciate if we could go over this in the morning and you can walk me through this stuff. I think I get everything but it'd help it stick in my brain if you'd walk me through the code.

File: features/step_definitions/cart_steps.rb (hunk 2 above)
Comment (tochman) [reply features/step_definitions/cart_steps.rb]: Yeah. Feel free to clean it up

File: app/controllers/carts_controller.rb (hunk 1 above)
Comment (tochman) [reply app/controllers/carts_controller.rb:18]: True

**Comment by tochman** [comment]
So if you update the description and clean up the code a bit this will be ready to be merged. I can walk you through my refactorings in the morning.

**Comment by AmberWilkie** [comment]
Ok np. Will do tomorrow morning.

**Review by tochman — COMMENTED** [review]
Do we have a test for hiding the Pay with Card button from not logged in users?

**Comment by AmberWilkie** [comment]
```
Scenario: Can only checkout as registered user
    Given there is one dish in my cart
    And I am on the "cart" page
    When I click the link "Log in to pay"
    Then I should be on the "register" page
    When I register as a user with username "Amber" and email "amber@random.com"
    Then I should be on the "cart" page
    And I should see "Pay with Card"
    And I should not see "Log in to pay"
```

Ok, that's only implicitly. I added a line about 'can't see Pay with Card'

File: app/controllers/carts_controller.rb [hunk 3]
```diff
@@ -11,17 +11,35 @@ def add_item
...
+    link_user_and_cart(params)
+    charge = StripePayment.perform_payment(params, @cart)
+    @order = ShoppingCart.find_by(user_id: current_user.id)
+    if charge.class == Stripe::Charge
+      mark_cart_as_paid(charge)
+      session.delete(:cart_id)
+      flash[:notice] = 'Your food is on its way!'
+      render :checkout
+    else
+      redirect_to carts_path, alert: charge
```
Comment (AmberWilkie) [inline app/controllers/carts_controller.rb:24]: What is this `alert: charge`? Is that going to return a pop-up box with a JSON string? Shouldn't it return something more useful (to the end user)?

Comment (tochman) [reply app/controllers/carts_controller.rb:24]: It's the error message. Not json, it's a string. There's a test for it.

Comment (AmberWilkie) [reply app/controllers/carts_controller.rb:24]: ok very well. So... should we get Raoul to look at this or is one of us going to do the dogfood thing?

**Pull request merged**